moto should not take over `responses` defaults

We are using moto to mock AWS services (namely S3); and responses to control our access to third party APIs.

However, moto unconditionnally configures a global pass-through in `responses`:
At import time, `moto` configures responses:
https://github.com/spulec/moto/blob/master/moto/core/models.py#L249-L252

Once moto has been imported, it is no longer possible to use `responses` in one's code.

Since moto uses `@mock_XXX` idioms, it should only apply its configuration on code that is wrapped by such decorators/context managers.

I encountered this with tests that use `responses` but not `moto`. The errors only appeared when the tests were run after a test that did use `moto`.

```
self = <responses.RequestsMock object at 0x108894460>
adapter = <requests.adapters.HTTPAdapter object at 0x108894d60>, request = <PreparedRequest [GET]>
kwargs = {'cert': None, 'proxies': OrderedDict(), 'stream': False, 'timeout': None, ...}

    def _on_request(self, adapter, request, **kwargs):
>       match, match_failed_reasons = self._find_match(request)
E       TypeError: cannot unpack non-iterable Response object
```

The fix in #3529 looks sensible. In the meantime, this workaround works for me:

```python
@pytest.fixture
def configure_ssm():
    original_responses_mock = responses._default_mock
    responses._default_mock = responses.RequestsMock(assert_all_requests_are_fired=False)
    with mock_ssm():
        # ssm setup goes here
        yield
    responses._default_mock = original_responses_mock
```

--- a/moto/apigateway/models.py
+++ b/moto/apigateway/models.py
@@ -41,6 +41,7 @@ from .exceptions import (
     RestAPINotFound,
     ModelNotFound,
 )
+from ..core.models import responses_mock
 
 STAGE_URL = "https://{api_id}.execute-api.{region_name}.amazonaws.com/{stage_name}"
 
@@ -565,7 +566,7 @@ class RestAPI(BaseModel):
         )
 
         for url in [stage_url_lower, stage_url_upper]:
-            responses._default_mock._matches.insert(
+            responses_mock._matches.insert(
                 0,
                 responses.CallbackResponse(
                     url=url,

--- a/moto/core/models.py
+++ b/moto/core/models.py
@@ -248,7 +248,7 @@ botocore_mock = responses.RequestsMock(
     target="botocore.vendored.requests.adapters.HTTPAdapter.send",
 )
 
-responses_mock = responses._default_mock
+responses_mock = responses.RequestsMock(assert_all_requests_are_fired=False)
 # Add passthrough to allow any other requests to work
 # Since this uses .startswith, it applies to http and https requests.
 responses_mock.add_passthru("http")

--- a/tests/test_apigateway/test_apigateway.py
+++ b/tests/test_apigateway/test_apigateway.py
@@ -8,9 +8,9 @@ import requests
 import sure  # noqa
 from botocore.exceptions import ClientError
 
-import responses
 from moto import mock_apigateway, mock_cognitoidp, settings
 from moto.core import ACCOUNT_ID
+from moto.core.models import responses_mock
 import pytest
 
 
@@ -1786,8 +1786,8 @@ def test_get_model_with_invalid_name():
 
 @mock_apigateway
 def test_http_proxying_integration():
-    responses.add(
-        responses.GET, "http://httpbin.org/robots.txt", body="a fake response"
+    responses_mock.add(
+        responses_mock.GET, "http://httpbin.org/robots.txt", body="a fake response"
     )
 
     region_name = "us-west-2"

--- a/tests/test_sns/test_publishing_boto3.py
+++ b/tests/test_sns/test_publishing_boto3.py
@@ -8,11 +8,11 @@ import re
 from freezegun import freeze_time
 import sure  # noqa
 
-import responses
 from botocore.exceptions import ClientError
 import pytest
 from moto import mock_sns, mock_sqs, settings
 from moto.core import ACCOUNT_ID
+from moto.core.models import responses_mock
 from moto.sns import sns_backend
 
 MESSAGE_FROM_SQS_TEMPLATE = (
@@ -332,7 +332,7 @@ def test_publish_to_http():
         json.loads.when.called_with(request.body.decode()).should_not.throw(Exception)
         return 200, {}, ""
 
-    responses.add_callback(
+    responses_mock.add_callback(
         method="POST", url="http://example.com/foobar", callback=callback
     )
 

